Added offline geocode sample

manishk3189 opened the pull request:
Hi @doneill I have created an offline geocoding/reverse geocoding sample. Can you please review the same and let me know of changes if required?

Thanks

doneill commented on offline-geocode/README.md:
```diff
@@ -0,0 +1,72 @@
+# How to use the sample
+The sample depends on basemap data to be located on the device. This includes installing a local tile map cache (tpk) to device as described below:
+
+ * Download San Francisco tpk and locator data from ArcGIS.
```
@manishk3189 Where do you download the data from?

doneill commented:
@manishk3189 Need to know where to get the data from?

manishk3189 commented:
@doneill I have added links in the README to download `tpk` and `locator` data from ArcGIS.

doneill commented:
@manishk3189 Are the find-route icons supposed to be included with this PR?  If not, please remove.

doneill commented on offline-geocode/README.md:
```diff
@@ -0,0 +1,72 @@
+  * /{device-externalstoragepath}/ArcGIS/samples/Offlinegeocode/Geocoding/SanFrancisco.loc
+  * /{device-externalstoragepath}/ArcGIS/samples/Offlinegeocode/Geocoding/SanFrancisco.locb
+  * /{device-externalstoragepath}/ArcGIS/samples/Offlinegeocode/Geocoding/SanFrancisco.lox
+  * /{device-externalstoragepath}/ArcGIS/samples/Offlinegeocode/Geocoding/SanFranciscoLocator.loc.xml
```
Remove the additional **/Geocoding/** folder as sample does not work when putting locator data in the folder, should be in the folder above

doneill commented:
@manishk3189 Here are some UI/UX comments to work through in addition to the comment on data folders above:  

Geocoding:
- Remove the ActionBar and replace with a Floating Search Bar
- Move the Search button to the left of the search bar to initiate the search, hitting enter on the soft keyboard should also initiate a search
- Create a down arrow button on the left of the search bar to show the list of examples

Something similar to the sample screenshot below:  

![pc](https://cloud.githubusercontent.com/assets/1308001/16933132/f439c4ba-4cff-11e6-8d80-6fe864a22514.png)


Reverse Geocoding: 
- Use long press to initiate a reverse geocoding search
- Tap hold and drag to allow for real time reverse geocoding.

manishk3189 commented:
@doneill I have refactored the sample with the following changes : 
1. Included the floating Search Bar.
2. The search bar has a dropdown on right which shows a list of address to geocode. 
3. In order to reverse geocode, long press on a location to get the result address in a callout. 
4. For real-time reverse geocoding, select the **Pin** marker symbol and perform a drag gesture anywhere on the map and results will be reverse geocoded on the fly. 
I have also updated the README with the instructions on how to perform geocoding and reverse geocoding.

Can you please review the PR and let me know of changes if required?

Thanks

manishk3189 commented:
@doneill The sample now performs real-time reverse geocoding without any touch delay between the position and screen Pin. Also, the address is updated as and when the pin is moved. I have also refactored the drop-down list to a position below the search bar and viewpoint when the sample is launched. Can you please review the changes and let me know of any improvements if needed.

Thanks

doneill commented on offline-geocode/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,22 @@
+    <string name="location_not_foud">We are sorry, we couldn\'t find a location with the name - </string>
+    <!-- San Francisco tpk and locator paths -->
+    <string name="sandiego_tpk">/ArcGIS/samples/OfflineRouting/streetmap_SD.tpk</string>
+    <string name="sandiego_loc">/ArcGIS/samples/OfflineRouting/Geocoding/SanDiego_StreetAddress.loc</string>
```
@manishk3189 The folder path on device does not match the folder path in the README.  Since this is the offline-geocoding sample use change the **/OfflineRouting/** to **/OfflineGeocoding/** for the device file path.

doneill replied:
@manishk3189 The locator file has not been updated so the app does not geocode.

manishk3189 commented:
@doneill I have updated the README and string resource with directory structure **/ArcGIS/samples/OfflineGeocoding/**. Also, I have removed extra **Geocoding** directory which was earlier used for locator data set.

doneill commented on offline-geocode/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,22 @@
+    <string name="location_not_foud">We are sorry, we couldn\'t find a location with the name - </string>
+    <!-- San Francisco tpk and locator paths -->
+    <string name="sandiego_tpk">/ArcGIS/samples/OfflineGeocoding/streetmap_SD.tpk</string>
+    <string name="sandiego_loc">/ArcGIS/samples/OfflineGeocoding/SanDiego_StreetAddress.loc</string>
```
@manishk3189 This is not the correct locator file so the app does not geocode.

manishk3189 commented:
@doneill Apologies, the string resource is updated now. I have also deleted the temporary folder containing the tpk from the PR

doneill commented on offline-geocode/src/main/res/layout/searchview.xml:
```diff
@@ -0,0 +1,31 @@
+    android:layout_marginLeft="15dp"
+    android:layout_marginRight="15dp"
+    android:layout_marginTop="20dp"
+    android:background="#ffff"
```
`@style/AppTheme` paints this background so I don't think you need this, test by removing.

manishk3189 replied:
@doneill I tested by removing the color background on LinerLayout and launched the app, the background was transparent. After that I tried to set a white background to both the  Spinner and SearchView, in this case, Spinner's dropdown icon was also painted white. Hence we will need to keep the background in LinerLayout.

doneill commented on offline-geocode/src/main/java/com/esri/arcgisruntime/sample/offlinegeocode/MainActivity.java:
```diff
@@ -0,0 +1,592 @@
+    private GeocodeParameters mGeocodeParameters;
+    private PictureMarkerSymbol mPinSourceSymbol;
+    private ArcGISMap mMap;
+    private ArcGISTiledLayer tiledLayer;
```
`ArcGISMap` & `ArcGISTiledLayer` can be converted to local variable.

doneill commented on offline-geocode/src/main/java/com/esri/arcgisruntime/sample/offlinegeocode/MainActivity.java:
```diff
@@ -0,0 +1,592 @@
+    private String mGraphicPointAddress;
+    private Point mGraphicPoint;
+    private GeocodeResult mGeocodedLocation;
+    private Spinner mSpinner;
```
`Spinner` can be local variable.

doneill commented on offline-geocode/src/main/java/com/esri/arcgisruntime/sample/offlinegeocode/MainActivity.java:
```diff
@@ -0,0 +1,592 @@
+    private static final String TAG = "OfflineActivity";
+    private final String extern = Environment.getExternalStorageDirectory().getPath();
+    int requestCode = 2;
+    String[] permission = new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE};
```
2 Fields above can be private

doneill commented on offline-geocode/src/main/java/com/esri/arcgisruntime/sample/offlinegeocode/MainActivity.java:
```diff
@@ -0,0 +1,592 @@
+    /**
+     * Hides soft keyboard
+     */
+    protected void hideKeyboard() {
```
Method should be private

doneill commented on offline-geocode/src/main/java/com/esri/arcgisruntime/sample/offlinegeocode/MainActivity.java:
```diff
@@ -0,0 +1,592 @@
+                        // get the list of graphics returned by identify
+                        List<Graphic> graphic = identifyGraphic.get();
+                        // get size of list in results
+                        int identifyResultSize = graphic.size();
```
`identifyResultSize` seems to be unused, see if you can remove.

doneill commented:
@manishk3189 Updates look great! Just some minor comments to address, then we should be able to merge!

manishk3189 commented:
@doneill I have accomodated the the review comments in the code. However, I haven't changed the removing of background resource for the **LinearLayout** in `searchview.xml` as removing it resulted in transparent background.

doneill commented:
Thanks @manishk3189 I will merge.  Nice work!
